File: src/main/kotlin/BJ1463.kt
import java.util.PriorityQueue

object BJ1463 {
    data class Cal(val num: Int, val count: Int) {
        fun divideBy3(): Cal = Cal(num / 3, count + 1)
        fun divideBy2(): Cal = Cal(num / 2, count + 1)
        fun minusOne(): Cal = Cal(num - 1, count + 1)
        fun isKeep(dp: IntArray): Boolean = dp[num] > count && num > 0
        fun nextCalList(): List<Cal> {
            val nextCals = arrayListOf<Cal>()
            if (num % 3 == 0) nextCals.add(divideBy3())
            if (num % 2 == 0) nextCals.add(divideBy2())
            if (num > 1) nextCals.add(minusOne())
            return nextCals
        }
    }
    fun IntArray.minDpUpdate(cal: Cal) {
        this[cal.num] = cal.count
    }
    fun solution(n: Int): Int {
        val queue: PriorityQueue<Cal> = PriorityQueue(compareBy { it.count })
        val dp = IntArray(n + 1) { Int.MAX_VALUE }
        queue.add(Cal(n, 0))
        dp[n] = 0

        while (queue.isNotEmpty()) {
            val cal = queue.poll()
            if (cal.num == 1) return cal.count
            if (cal.isKeep(dp)) continue
            val cals = cal.nextCalList()
            for (c in cals) {
                if (c.isKeep(dp)) {
                    dp.minDpUpdate(c)
                    queue.add(c)
                }
            }
        }

        return n - 1
    }

    fun solve() {
        val n = readln().toInt()
        println(solution(n))
    }

    @JvmStatic
    fun main(args: Array<String>) {
        solve()
    }
}

File: python/src/bj1463.py
from collections import deque


def solution(start: int) -> int:
    q = deque()
    q.append([start, 0])
    count = 0

    while len(q) > 0:
        num, c = q.popleft()
        if num == 1:
            count = c
            break
        if num % 3 == 0:
            q.append([num // 3, c + 1])
        if num % 2 == 0:
            q.append([num // 2, c + 1])
        q.append([num - 1, c + 1])

    return count


if __name__ == "__main__":
    x = int(input())
    print(solution(start=x))
